query.go

```diff
@@ -103,14 +122,339 @@ func DoneChan(doneChan chan<- struct{}) QueryOption {
+		// and removing disconnected/banned peers.
+		for _, peer := range s.Peers() {
+			if _, ok := peerQuits[peer]; !ok && peer.Connected() {
+				peerQuits[peer] = make(chan struct{})
+				matchSignals[peer] = make(chan wire.Message)
+				go peerGoroutine(peer, peerQuits[peer],
+					matchSignals[peer])
+			}
+
+		}
+
+		for peer, quitChan := range peerQuits {
+			if !peer.Connected() {
+				close(quitChan)
+				close(matchSignals[peer])
+				delete(peerQuits, peer)
+				delete(matchSignals, peer)
+			}
+		}
+	}
+
+}
+
+// queryAllPeers is a helper function that sends a query to all peers and waits
+// for a timeout specified by the QueryTimeout package-level variable or the
+// Timeout functional option. The NumRetries option is set to 1 by default
+// unless overridden by the caller.
+func (s *ChainService) queryAllPeers(
+	// queryMsg is the message to broadcast to all peers.
+	queryMsg wire.Message,
+
+	// checkResponse is called for every message within the timeout period.
+	// The quit channel lets the query know to terminate because the
+	// required response has been found. This is done by closing the
+	// channel. The peerQuit lets the query know to terminate the query for
+	// the peer which sent the response, allowing releasing resources for
+	// peers which respond quickly while continuing to wait for slower
+	// peers to respond and nonresponsive peers to time out.
+	checkResponse func(sp *ServerPeer, resp wire.Message,
+		quit chan<- struct{}, peerQuit chan<- struct{}),
+
+	// options takes functional options for executing the query.
+	options ...QueryOption) {
+
+	// Starting with the set of default options, we'll apply any specified
+	// functional options to the query.
+	qo := defaultQueryOptions()
+	qo.numRetries = 1
+	for _, option := range options {
+		option(qo)
+	}
+
+	// This is done in a single-threaded query because the peerState is
+	// held in a single thread. This is the only part of the query
+	// framework that requires access to peerState, so it's done once per
+	// query.
+	peers := s.Peers()
+
+	// This will be shared state between the per-peer goroutines.
+	quit := make(chan struct{})
+	allQuit := make(chan struct{})
+	var wg sync.WaitGroup
+	msgChan := make(chan spMsg)
+	subscription := spMsgSubscription{
+		msgChan:  msgChan,
+		quitChan: allQuit,
+	}
+
+	// Now we start a goroutine for each peer which manages the peer's
+	// message subscription.
+	peerQuits := make(map[*ServerPeer]chan struct{})
+	for _, sp := range peers {
+		sp.subscribeRecvMsg(subscription)
+		wg.Add(1)
+		peerQuits[sp] = make(chan struct{})
+		go func(sp *ServerPeer, peerQuit <-chan struct{}) {
+			defer wg.Done()
+
+			defer sp.unsubscribeRecvMsgs(subscription)
+
+			for i := uint8(0); i < qo.numRetries; i++ {
+				timeout := time.After(qo.timeout)
+				sp.QueueMessageWithEncoding(queryMsg,
+					nil, qo.encoding)
+				select {
+				case <-quit:
+					return
+				case <-peerQuit:
+					return
+				case <-timeout:
```

Comment:
What use is the timeout here? I guess it should be the case that the peer will exit if the caller detects that the correct response was received?